### PR TITLE
Fix generic recursion condition when calling DeserializeAsyncEnumerable in NativeAOT.

### LINKED ISSUE #84922: 8.4 MB size regression updating System.Text.Json to 8.0.0
### Description

Now that #84895 has been fixed, I was able to start dogfooding System.Text.Json 8.0.0 builds in the Microsoft Store, to gather some numbers so we can more easily track the binary size difference with this new release. This is where I've noticed an problem: my understanding is that the new 8.0.0 release is expected to include several binary size improvements, which should make the update result in a _smaller_ binary size footprint (see #79122), but I'm actually seeing a regression, and not even a trivial one 🥲

### Configuration

| System.Text.Json version | Microsoft Store package size |
| --- | --- |
| 7.0.2 |  66.410 KB |
| 8.0.0-preview.4.23216.5 | **70.015 KB** |

Delta: **+3.605 KB** (**+5.4%**). The real delta for the x64 binary is actually ~8.4 MB (ie. ~17%), see details below.

### Regression?

Yes. The binary size is actually, surprisingly, lower on 7.0.2.

cc. @eiriktsarpalis @layomia

I assume there's definitely something wrong going on here?
Ie. my assumption would be that updating to 8.0.0 should definitely result in a net _decrease_ of binary size, no? 🤔

Happy to help test any preview bits necessary and/or to provide more info (internally if needed, as the project is not public).

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
This PR makes the following changes:

1. Moves the `JsonTypeInfo<Queue<T>>` factory from `JsonTypeInfo<T>` to `JsonSerializer` to avoid triggering polymorphic recursion.
2. Removes the non-generic `DeserializeAsyncEnumerable` APIs added in Preview 1 via #79397. Supporting that feature necessitates every `JsonTypeInfo<T>` encapsulating the factory for `JsonTypeInfo<Queue<T>>` triggering the recursion condition.

I've validated that the AOT warning no longer occurs using manual testing.

Fix https://github.com/dotnet/runtime/issues/84922.

cc @Sergio0694

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs:473]: @jkotas @MichalStrehovsky after a lot of experimentation, I was able to pin down the AOT recursion warnings to this method being defined within `JsonTypeInfo<T>` itself, regardless of whether it's static or not. I find that somewhat surprising, any chance we could be looking at a false positive in the AOT compiler?

**Comment by Sergio0694** [comment]
Thank you!! Will validate with the Store too and report back 🙂

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs
Comment (jkoritzinsky) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:250]: Add a <kbd>breaking-change</kbd> label to this PR and fill out a breaking change doc?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:250]: This is API added in .NET 8 Preview 1. Do we need the breaking-change process for API changes in the current version?

Comment (jkoritzinsky) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:250]: I think across some preview boundaries we've done the whole process. Usually if we've announced the API in a blog post or at //BUILD or something like that we follow the process. Not sure what we do otherwise.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:250]: This particular set of APIs wasn't announced anywhere, since it's part of the ongoing Goldilocks/NativeAOT effort. These APIs have been fluctuating a lot across previews anyway.

Comment (jkoritzinsky) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:250]: Yeah then we can skip the process.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs
Comment (jkotas) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs:473]: It is limitation of AOT compiler analysis. Here is minimal repro:
```
Console.WriteLine(new G<string>().M());

class G<T>
{
    public object M()
    {
        return new G<List<T>>();
    }
}
```

When the `G<List<T>>` instance is created in `M()`, we assume that it is possible for `M()` to be called on that instance again, and that triggers the infinite generic recursion. I have opened https://github.com/dotnet/runtime/issues/85184 on this.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.ReadHelper.cs
Comment (jkotas) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.ReadHelper.cs:84]:
```suggestion
        internal JsonTypeInfo<Queue<T>>? _asyncEnumerableQueueTypeInfo;
```
Nit: Naming convention for fields.

**Comment by jkotas** [comment]
LGTM once @Sergio0694 confirms that it fixes the problem for store. Thank you!

**Review by jkotas — APPROVED** [review]

**Comment by MichalPetryka** [comment]
> LGTM once @Sergio0694 confirms that it fixes the problem for store. Thank you!

https://github.com/dotnet/runtime/issues/84922#issuecomment-1518352863

**Review by Sergio0694 — APPROVED** [review]
Just commented in the other thread, LGTM!! :shipit:

@eiriktsarpalis thank you!! 🙌

**Review by stephentoub — APPROVED** [review]
Other than the pending feedback, LGTM

**Pull request merged**
